otherlibs/dune-rpc/private/versioned.ml

```diff
@@ -0,0 +1,434 @@
+
+       from the type-erasure implementation shenanigans described above, letting
+       all related operations (declaring, implementing, dispatching) share
+       uniform implementations as much as possible. *)
+    type (_, _, _) field_witness =
+      | Declared_requests
+          : ( _
+            , string * ('req, 'resp) Decl.Generation.t Int.Map.t Univ_map.Key.t
+            , ('req, 'resp) Decl.Generation.t )
+            field_witness
+      | Declared_notifs
+          : ( _
+            , string * ('a, unit) Decl.Generation.t Int.Map.t Univ_map.Key.t
+            , ('a, unit) Decl.Generation.t )
+            field_witness
+      | Impl_requests : ('state, string, 'state r_handler) field_witness
+      | Impl_notifs : ('state, string, 'state n_handler) field_witness
+
+    let get (type st a b) (t : st t) (witness : (st, a, b) field_witness)
+        (key : a) : b Int.Map.t option =
+      match witness with
+      | Declared_requests ->
+        let _, key = key in
+        let _, table = t.declared_requests in
+        Univ_map.find table key
+      | Declared_notifs ->
+        let _, key = key in
+        let _, table = t.declared_notifications in
+        Univ_map.find table key
+      | Impl_requests -> String.Table.find t.implemented_requests key
+      | Impl_notifs -> String.Table.find t.implemented_notifications key
+
+    let set (type st a b) (t : st t) (witness : (st, a, b) field_witness)
+        (key : a) (value : b Int.Map.t) =
+      match witness with
+      | Declared_requests ->
+        let name, key = key in
+        let known_keys, table = t.declared_requests in
+        t.declared_requests <-
+          ( String.Map.add_multi known_keys name (T key)
+          , Univ_map.set table key value )
+      | Declared_notifs ->
+        let name, key = key in
+        let known_keys, table = t.declared_notifications in
+        t.declared_notifications <-
+          ( String.Map.add_multi known_keys name (T key)
+          , Univ_map.set table key value )
+      | Impl_requests -> String.Table.set t.implemented_requests key value
+      | Impl_notifs -> String.Table.set t.implemented_notifications key value
+
+    let registered_procedures
+        { declared_requests = declared_request_keys, declared_request_table
+        ; declared_notifications =
+            declared_notification_keys, declared_notification_table
+        ; implemented_requests
+        ; implemented_notifications
+        } =
+      let batch_declarations which declared_keys declaration_table =
+        String.Map.foldi declared_keys ~init:[] ~f:(fun name keys acc ->
+            let generations =
+              List.fold_left keys ~init:[] ~f:(fun acc (T key) ->
+                  match Univ_map.find declaration_table key with
+                  | None ->
+                    Code_error.raise
+                      ("versioning: method in [known_" ^ which
+                     ^ "_table] without actually being declared")
```

Comment:
The message in a code error should always be constant. It makes it easier to group the code errors by this field. You should add the table name in a parameter if you want to include it.